### DISCUSSION TITLE
Dedicated turning vs. dual carriageways

### DISCUSSION THREAD

**Discussion by originalfoo** [CitiesSkylinesMods/TMPE#1254] [Ideas]
@kianzarrin Currently the dedicated turning lanes does this (which is what it's designed to do):

![image](https://user-images.githubusercontent.com/1386719/147229126-4e738355-899b-471f-866e-804da1c2bc3f.png)

However, in the case of [dual carriageways](https://en.wikipedia.org/wiki/Dual_carriageway) (DCs), normally there wouldn't be dedicated turning lanes on the major road _unless it's asymmetric_.

Example of a Spanish DC (image from wikipedia): 
![image](https://user-images.githubusercontent.com/1386719/147229617-0da49830-5b5a-49fc-9ff2-6880739dc610.png)

The outer lanes are generally used by slower vehicles (lorries/trucks) and they expect to be able to drive the whole route without needing to enter the inner lanes (except for occasional overtaking).

Example of a DC in Brazil (image from wikipedia):

![image](https://user-images.githubusercontent.com/1386719/147230000-5d15738b-6a22-42da-a789-ff2332d052e8.png)

In-game, the dedicated turning lanes on the DC are causing strange traffic weaving behaviour at junctions. Normally traffic would stay in lane, but because one lane is being "terminated" (due to dedicated turning lane) traffic is having to weave in to the outer/inner (as applicable) lane as it crosses the junction then back in to its original lane after the junction.

Example of a DC in UK - note the far-side turning lane available due to asymmetry:

![image](https://user-images.githubusercontent.com/1386719/147231725-4f5a3355-bc68-494e-9166-4e7da70c0f71.png)

The issue is also exacerbated when the junction is on a "kink" in the DC (image below); it's particularly obvious when in first person cam driving down such roads as the vehicles make startling moves when encountering such junctions.

![image](https://user-images.githubusercontent.com/1386719/147231042-7aeb577d-123b-4110-9a43-6d24f3bfe166.png)

Ideally there shouldn't be dedicaed turning lanes on the DC (a 2+2 lane road in general, not just those with medians) except in following cases:

1. Asymmetric road (the extra lane can be used as dedicated turning lane)
2. 3+ lanes per direction (as this allows slow trucks to use middle lane to bypass any traffic waiting in outer turning lane)

Would it be possible to get an option (eg. "Avoid turning lanes in 2+2 medium roads") to facilitate this? Otherwise the dedicated turning lane feature is making dual carraige ways cumbersome to work with b/c I now have to adjust the lane arrows all the way down those - usually very long - roads.

The result would look like this:

![image](https://user-images.githubusercontent.com/1386719/147233250-3463bbe1-9c0f-4e04-a9e5-fe4c887a7980.png)

Note also that I can't use `Alt+Click` or `Ctrl+Click` shortcuts in Lane Arrows tool to toggle turning lanes (is that a bug btw?); even if I could, it would still be PITA to have to constantly be doing that when working with DC roads.

I know we're in code freeze, but hopefully this wouldn't be too painful to implement because it's just checking for 2+2 lanes before applying turning lanes; pseudocode: `if numForward = 2 and numBackward = 2 then ignore`.

**Comment by kianzarrin** [comment]
![image](https://user-images.githubusercontent.com/26344691/147249074-4ee2af58-fb6d-47c6-994c-b8d9ac34a432.png)
what about the the far turn in the image above? should that be dedicated?

**Comment by kianzarrin** [comment]
in this pic there is a median. 
![image](https://user-images.githubusercontent.com/26344691/147249196-29492deb-e39b-4494-a3a3-d4e549eabbdb.png)

TMPE does this to high priority roads.

**Reply by originalfoo** [reply]
> what about the the far turn in the image above? should that be dedicated?

Only if there's an extra lane added (see the google image I posted), or the road is 3+3.

When you have outer or inner lane suddenly becoming an turning lane, any traffic wanting to go straight on now has to move in to the other lane to bypass the turning lane, so they are also disrupting the other lane. At those points, the 2+2 road becomes a turning lane + traffic accident/jam lane. It would be much better, in that situation, to have a 1+1 road, and occasionally extend it with extra lane(s) to use as turning lanes at junctions; the traffic flow would likely be higher than a 2+2 road with "turning lane choke points" along it.

The specific place where turning lanes shouldn't be added is T-junction with smaller road (as shown here: https://github.com/CitiesSkylinesMods/TMPE/discussions/1254#discussioncomment-1863551 ). If user wants dedicated turning lane on the main road at that junction, they should upgrade to a 2+3 asymmetric road.

Obviously, if there are traffic lights (ie. major junction) then dedicated turning lanes would make sense as the traffic is already slowed for the lights.

**Reply by kianzarrin** [reply]
What if there are allies joining from left and right to a 3(in each direction) lane road?

**Reply by originalfoo** [reply]
They would get turning lanes.

**Reply by kianzarrin** [reply]
I feel like this fits into another option like Improved lane arrows. there is already a mod for that but we can also bring that into TMPE.

**Reply by originalfoo** [reply]
I thought the improved lane arrows mod does essentially same as what our dedicated turning lanes option does?

**Reply by kianzarrin** [reply]
dedicated turning lane is just that. Its different than improving lane arrows.
if a city is gridlock this might be a quick fix. but might not be the best fix.
To improve lane arrows we need another policy.

**Reply by originalfoo** [reply]
Something like this?

```
(o) Dedicated turning lanes // the existing option
    (o) Avoid turning lanes on 2+2 roads
```

**Reply by kianzarrin** [reply]
Something like this:
(o) Dedicated turning lanes // the existing option
(o) Improved turning lanes // steal code from [existing mod](https://steamcommunity.com/sharedfiles/filedetails/?id=2119310640)

**Reply by kianzarrin** [reply]
but only one of them should be selected. so maybe something like this:
- improve lane arrows
    - Dedicated turning lanes 
    - Distributed turning lanes

**Comment by originalfoo** [comment]
Stem user just posted this:

![image](https://user-images.githubusercontent.com/1386719/153065805-0811d6c5-ac20-41bf-8428-ec27c6ff501c.png)

Which makes me think maybe the option should be something like:

```
On suitable roads, add dedicated turning lanes for:
    (o) Far-side turns (inner-lane; cutting across oncoming traffic)
    (o) Near-side turns (outer-lane; similar to a highway exit)
    (o) Only add near-side turning lanes if there are 3+ lanes
```

The `Limit...` option would _not_ apply to 2+ lane roads where there is no "straight ahead" route.

**Reply by bailsafe** [reply]
Yep, that was me! My preferred lane configuration, which would be slightly truer to life, would be something like this:

![TMPE Turn 1](https://user-images.githubusercontent.com/30294655/153076879-e791d207-2d69-4072-a214-6d5bd2a04519.png)

Of course, this would leave the inside lane technically unassigned (though I suppose it could be exclusive to the far-side turn from the bottom). This likely wouldn't happen in real life as you'd probably have the dedicated far-side turn [terminate ahead](https://user-images.githubusercontent.com/30294655/153077747-23d4b1d3-a519-489b-b691-96f5dbae8a74.png) to prevent ongoing traffic from using the lane, but that becomes an issue at T intersections, where [ongoing traffic on the near side must swerve](https://user-images.githubusercontent.com/30294655/153078515-1d21e6ef-8662-4dbd-8276-1bc7d335ec5c.png) to avoid the median.

Edit: Even with a 2+3 asymmetrical road, the default behavior seems odd in that it leaves the near-side lane without straight traffic.

![image](https://user-images.githubusercontent.com/30294655/153081004-71babe18-d9bf-4a10-a7f1-c1c9dfbd8b4f.png)

**Reply by originalfoo** [reply]
If the lane exists ahead, you'd still need way for traffic to reach it, such as this:

![image](https://user-images.githubusercontent.com/1386719/153080861-b6c52ba2-ae3d-4e07-b18c-177621aa61a0.png)

Ideally though you'd want 2+2 road for most of the route, but occasionally switch to 3+2 segment where a far-side turning lane is required.

**Reply by bailsafe** [reply]
Right; even in this case, though, you wouldn't normally have a near-side turning lane except in very special cases. It would only leave one lane going on to a two-lane segment. Using the same asymmetrical segment to make it 2-straight/1-near-turn would create the swerving traffic issue though.

![image](https://user-images.githubusercontent.com/30294655/153088827-6dfff9b6-e10d-49ac-94a8-b58f31411522.png)

Side note: The other issue with 2+2 to 3+2 is the lack of networks that create a [smooth connection between the two](https://user-images.githubusercontent.com/30294655/153089473-5e33bc73-75c5-43b1-ae95-99c44f4b068e.png), even with IMT as in the image, further diminishing realism 😅

**Reply by originalfoo** [reply]
> It would only leave one lane going on to a two-lane segment.

Yup, lane math needs considering when doing turning lanes. I don't think we've thought of it from that angle before.

> Side note: The other issue with 2+2 to 3+2 is the lack of networks that create a smooth connection between the two ...

Yup, I spend ages with Node Controller Renewal and IMT tweaking such junctions.

**Reply by bailsafe** [reply]
I think it would be beneficial to have the options list as you proposed:

```
On suitable roads, add dedicated turning lanes for:
    ○ Far-side turns (crosses opposing traffic)
    ○ Near-side turns (nearest to outer lane)
        ○ Only if road is 3+ lanes in one direction
```

It would help to define "suitable roads", though. Will it include both single and dual carriageways? Will it make exceptions for bus lanes?

**Reply by originalfoo** [reply]
> It would help to define "suitable roads", though. Will it include both single and dual carriageways? Will it make exceptions for bus lanes?

Yeah, there's only so much info we can covey via the options screen. There are loads of edge cases (bus lanes are treated as dedicated right of way for buses so aren't counted as road traffic lanes IIRC - if user needed extra lane as turning lane, they'd have to replace segment at junction with one that doesn't have bus lanes to allow other traffic to use it).

IMO we could probably do with an `Only if road is 3+ lanes` sub-option for the far-side turns too? In many cases IRL (at least in Europe) 2+2 roads don't get dedicated turning lanes - they'd usually add a 3+2 "segment" near traffic lights to facilitate dedicated lane for far-side turns.

**Reply by kianzarrin** [reply]
There is a difference between dedicated turning lane and improved lane arrows. There is a mod that does the latter. Maybe we should integrate it inside TMPE?

**Reply by kianzarrin** [reply]
If somebody asked for improved turning lane just refer the to that mod.

**Comment by wzhec** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。
